### PR TITLE
Extension kubernetes registry

### PR DESCRIPTION
**Pull request by wangchengming666**
# 总体设计

image

# 样例
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
   com.alipay.sofa.rpc.registry.kubernetes.TestService:1.0@DEFAULT: 'bolt://192.168.1.100:12200?accepts=100000&appName=TestAppName&weight=100&language=java&pid=41517&interface=com.alipay.sofa.rpc.registry.kubernetes.TestService&timeout=0&serialization=hessian2&protocol=bolt&delay=20&dynamic=true&startTime=1711021435028&id=rpc-cfg-0&uniqueId=&rpcVer=51300'
  creationTimestamp: "2024-02-02T08:47:54Z"
  generateName: sofa-rpc-k8s-provider-86f8565dcb-
  labels:
    com.alipay.sofa.rpc.registry.kubernetes.TestService:1.0@DEFAULT: ''
    app: sofa-rpc-k8s-provider
    pod-template-hash: 86f8565dcb
  name: sofa-rpc-k8s-provider-86f8565dcb-m7r6v
  namespace: sofa-rpc-k8s
```

# 环境准备
1. 安装docker
https://www.docker.com/
2. 安装minikube
https://kubernetes.io/zh-cn/docs/tutorials/hello-minikube/

# 代码准备
https://github.com/wangchengming666/sofa-rpc-registry-kubernetes-demo

# 测试手法
## Provider部分
1. 打包provider的镜像
`docker build -t sofa-rpc-k8s-provider:1.0.0 .`
2. 加载镜像
`minikube image load sofa-rpc-k8s-provider:1.0.0  `
3. apply sa
`kubectl apply -f ServiceAccount.yml  `
4. apply Deployment
`kubectl apply -f Deployment.yml  `
5. 查看provider的pod
`kubectl get pods -n sofa-rpc-k8s `

```
chengming@chengmingdeMacBook-Pro k8s % kubectl get pods -n sofa-rpc-k8s                                          
NAME                                     READY   STATUS    RESTARTS   AGE
sofa-rpc-k8s-provider-86f8565dcb-9z7hw   1/1     Running   0          7s
sofa-rpc-k8s-provider-86f8565dcb-jw4wg   1/1     Running   0          7s
sofa-rpc-k8s-provider-86f8565dcb-m7r6v   1/1     Running   0          7s
```
6. 查看provider的日志
`kubectl logs -f sofa-rpc-k8s-provider-86f8565dcb-m7r6v  -n sofa-rpc-k8s  `

```
chengming@chengmingdeMacBook-Pro k8s % kubectl logs -f sofa-rpc-k8s-provider-86f8565dcb-m7r6v  -n sofa-rpc-k8s    

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::               (v2.4.13)

2024-02-02 08:48:02.078  INFO 1 --- [           main] com.example.appboot.AppbootApplication   : Starting AppbootApplication v0.0.1-SNAPSHOT using Java 1.8.0_342 on sofa-rpc-k8s-provider-86f8565dcb-m7r6v with PID 1 (/app/appboot-0.0.1-SNAPSHOT.jar started by root in /app)
2024-02-02 08:48:02.098  INFO 1 --- [           main] com.example.appboot.AppbootApplication   : No active profile set, falling back to default profiles: default
2024-02-02 08:48:12.561  INFO 1 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8080 (http)
2024-02-02 08:48:12.753  INFO 1 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
2024-02-02 08:48:12.757  INFO 1 --- [           main] org.apache.catalina.core.StandardEngine  : Starting Servlet engine: [Apache Tomcat/9.0.55]
2024-02-02 08:48:13.470  INFO 1 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/myapp]  : Initializing Spring embedded WebApplicationContext
2024-02-02 08:48:13.470  INFO 1 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 10590 ms
2024-02-02 08:48:17.359  INFO 1 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8080 (http) with context path '/myapp'
2024-02-02 08:48:17.380  INFO 1 --- [           main] com.example.appboot.AppbootApplication   : Started AppbootApplication in 18.826 seconds (JVM running for 21.29)
AppbootApplication start
AppbootApplication start success

```

## Consumer部分
1. 打包consumer的镜像
` docker build -t sofa-rpc-k8s-consumer:1.0.0 .`
2. 加载镜像
`minikube image load sofa-rpc-k8s-consumer:1.0.0  `
3. apply Deployment
`kubectl apply -f Deployment.yml  `
4. 查看consumer的pod
`kubectl get pods -n sofa-rpc-k8s `

```
chengming@chengmingdeMacBook-Pro k8s % kubectl get pods -n sofa-rpc-k8s                                                      
NAME                                     READY   STATUS    RESTARTS   AGE
sofa-rpc-k8s-consumer-5d6cd8698d-pf7wd   1/1     Running   0          4s
```
5. 查看consumer的日志
`kubectl logs -f sofa-rpc-k8s-consumer-5d6cd8698d-pf7wd -n sofa-rpc-k8s `

```
chengming@chengmingdeMacBook-Pro k8s % kubectl logs -f sofa-rpc-k8s-consumer-5d6cd8698d-pf7wd -n sofa-rpc-k8s   
Listening for transport dt_socket at address: 5005

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::               (v2.4.13)

2024-02-02 08:58:31.304  INFO 1 --- [           main] c.s.e.s.SofaRpcK8sConsumerApplication    : Starting SofaRpcK8sConsumerApplication v0.0.1-SNAPSHOT using Java 1.8.0_342 on sofa-rpc-k8s-consumer-5d6cd8698d-pf7wd with PID 1 (/app/sofa-rpc-k8s-consumer-0.0.1-SNAPSHOT.jar started by root in /app)
2024-02-02 08:58:31.308  INFO 1 --- [           main] c.s.e.s.SofaRpcK8sConsumerApplication    : No active profile set, falling back to default profiles: default
2024-02-02 08:58:34.300  INFO 1 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8082 (http)
2024-02-02 08:58:34.320  INFO 1 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
2024-02-02 08:58:34.320  INFO 1 --- [           main] org.apache.catalina.core.StandardEngine  : Starting Servlet engine: [Apache Tomcat/9.0.55]
2024-02-02 08:58:34.590  INFO 1 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/myapp]  : Initializing Spring embedded WebApplicationContext
2024-02-02 08:58:34.590  INFO 1 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 3165 ms
2024-02-02 08:58:36.019  INFO 1 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8082 (http) with context path '/myapp'
2024-02-02 08:58:36.077  INFO 1 --- [           main] c.s.e.s.SofaRpcK8sConsumerApplication    : Started SofaRpcK8sConsumerApplication in 5.676 seconds (JVM running for 6.578)
```
6. 发起调用，在控制台可以看到如下日志，代表调用成功
```
start call sofa-rpc-k8s-provider
2024-02-04 02:12:40.364  WARN 1 --- [           main] io.fury.config.FuryBuilder               : Class registration isn't forced, unknown classes can be deserialized. If the environment isn't secure, please enable class registration by `FuryBuilder#requireClassRegistration(true)` or configure ClassChecker by `ClassResolver#setClassChecker`
2024-02-04 02:12:41.593  INFO 1 --- [           main] io.fury.Fury                             : Created new fury io.fury.Fury@4d02f94e
2024-02-04 02:12:41.675  WARN 1 --- [           main] .c.h.i.AbstractStringBuilderDeserializer : coder field not found or not accessible, will skip coder check, error is coder
call provider success chengming
call provider success chengming
call provider success chengming
call provider success chengming
call provider success chengming
call provider success chengming
call provider success chengming
call provider success chengming
call provider success chengming
call provider success chengming
call provider success chengming
call provider success chengming
```

## Summary by CodeRabbit

- **New Features**
	- Introduced Kubernetes registry support for service registration and discovery.
	- Added utility classes for Kubernetes client configuration and interaction.
	- Enhanced `RegistryConfig` with methods to retrieve parameters with default values.
- **Tests**
	- Added tests for Kubernetes registry functionality, including service registration and discovery.
- **Documentation**
	- Added documentation entries for Kubernetes registry support.

### CODE REVIEW

**Comment by nobodyiam** [comment]
Looks interesting. Would you please briefly describe the behavior of the register and unRegister? It seems the information is written to the pod's annotation and label?

**Comment by wangchengming666** [comment]
> Looks interesting. Would you please briefly describe the behavior of the register and unRegister? It seems the information is written to the pod's annotation and label?

sure, I will add the doc on this feature

**Comment by wangchengming666** [comment]
related https://github.com/sofastack/sofa-boot/pull/1288

**Comment by Lihus** [comment]
朱立焕收到，谢谢

**Review by EvenLjj — CHANGES_REQUESTED** [review]

File: registry/registry-kubernetes/src/test/java/com/alipay/sofa/rpc/registry/kubernetes/KubernetesRegistryTest.java [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+                .setApplication(applicationConfig)
+                .setInterfaceId(TestService.class.getName())
+                .setRegistry(registryConfig)
+                .setRegister(true)
+                //                .setUniqueId("standalone")
+                .setRef(new TestServiceImpl())
+                .setServer(serverConfig);
+
+        // 注册
+        kubernetesRegistry.register(providerConfig);
```
Comment (EvenLjj) [inline registry/registry-kubernetes/src/test/java/com/alipay/sofa/rpc/registry/kubernetes/KubernetesRegistryTest.java:182]: TestCase 发布多个不同接口服务的case

File: registry/registry-kubernetes/src/main/java/com/alipay/sofa/rpc/registry/kubernetes/KubernetesRegistry.java [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    private ConcurrentMap<ProviderConfig, Pod> providerInstances = new ConcurrentHashMap<>(64);
+
+    private ConcurrentMap<ConsumerConfig, SharedIndexInformer<Pod>> consumerListeners = new ConcurrentHashMap<>(64);
+
+    /**
+     * Instantiates a new kubernetes registry.
+     *
+     * @param registryConfig
+     */
+    protected KubernetesRegistry(RegistryConfig registryConfig) {
```
Comment (EvenLjj) [inline registry/registry-kubernetes/src/main/java/com/alipay/sofa/rpc/registry/kubernetes/KubernetesRegistry.java:77]: 这里是否可以改成public

File: registry/registry-kubernetes/src/main/java/com/alipay/sofa/rpc/registry/kubernetes/KubernetesRegistry.java [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+                    .withName(currentHostname);
+
+            if (CommonUtils.isEmpty(providerInstances)) {
+                podResource.edit(pod -> new PodBuilder(pod).editOrNewMetadata()
+                        .removeFromAnnotations(ANNOTATION_KEY)
+                        .endMetadata()
+                        .build());
+            } else {
+                // 考虑可能会有一个pod上多个ProviderConfig的场景，所以先全部移除然后将最新的providerInstances存在Annotations上
+                podResource.edit(pod -> new PodBuilder(pod).editOrNewMetadata()
```
Comment (EvenLjj) [inline registry/registry-kubernetes/src/main/java/com/alipay/sofa/rpc/registry/kubernetes/KubernetesRegistry.java:164]: 同样，如果做成接口维度，unregister的时候就不会这么复杂。

**Review by EvenLjj — APPROVED** [review]
LGTM

**Review by nobodyiam — APPROVED** [review]
LGTM

**Pull request merged**